clash-protocols-memmap/src/Protocols/MemoryMap.hs

```diff
@@ -0,0 +1,340 @@
+
+  1. Provide a way for memory mapped peripherals to define their memory layout
+
+  2. Make it possible for designers to define hardware designs without specifying
+     exact memory addresses if they can be mapped to arbitrary addresses, while
+     not giving up the ability to.
+
+  3. Provide a way to check whether memory maps are valid, i.e. whether
+     perhipherals do not overlap or exceed their allocated space.
+
+  4. Provide a way to generate human readable documentation
+
+  5. Provide a way to generate data structures for target languages. I.e., if
+     a designer instantiates a memory mapped register on type @a@, it should
+     be possible to generate an equivalent data structure in Rust or C.
+-}
+type MM = SimOnly MemoryMap
+
+data ConstF (a :: Type)
+
+data ConstB (a :: Type)
+
+instance Protocol (ConstF a) where
+  type Fwd (ConstF a) = a
+  type Bwd (ConstF a) = ()
+
+instance Protocol (ConstB a) where
+  type Fwd (ConstB a) = ()
+  type Bwd (ConstB a) = a
+
+instance IdleCircuit (ConstB a) where
+  -- idleFwd :: Data.Proxy.Proxy (ConstB a) -> Fwd (ConstB a)
+  idleFwd Proxy = ()
+
+  -- idleBwd :: Data.Proxy.Proxy (ConstB a) -> Bwd (ConstB a)
+  idleBwd Proxy = error ""
+
+constF :: a -> Circuit () (ConstF a)
+constF val = Circuit $ \((), ()) -> ((), val)
+
+constB :: a -> Circuit (ConstB a) ()
+constB val = Circuit $ \((), ()) -> (val, ())
+
+getConstB :: Circuit (ConstB a) () -> a
+getConstB (Circuit f) = fst $ f ((), ())
+
+data Name = Name
+  { name :: String
+  -- ^ Name of the 'thing'. Used as an identifier in generated data structures.
+  --
+  -- TODO: Only allow very basic names here to make sure it can be used in all
+  --       common target languages. Provide a Template Haskell helper to make
+  --       sure the name is valid at compile time?
+  , description :: String
+  -- ^ Description of the 'thing'. Used in generated documentation.
+  }
+  deriving (Show, Eq, Ord)
+
+{- | Wrapper for \"things\" that have a name and a description. These are used
+to generate documentation and data structures for target languages.
+-}
+
+-- type Named a = (Name, a)
+
+type NamedLoc a = (Name, SrcLoc, a)
+
+type DeviceDefinitions = Map.Map DeviceName DeviceDefinition
+
+data MemoryMap = MemoryMap
+  { deviceDefs :: DeviceDefinitions
+  , tree :: MemoryMapTree
+  }
+  deriving (Show)
+
+data DeviceDefinition = DeviceDefinition
+  { deviceName :: Name
+  , registers :: [NamedLoc Register]
+  , defLocation :: SrcLoc
+  , tags :: [String]
+  }
+  deriving (Show)
+
+deviceSize :: DeviceDefinition -> Integer
+deviceSize def' =
+  List.foldr
+    ( \(_, _, reg) acc ->
+        max acc (reg.address + regByteSizeC reg.fieldType)
+    )
+    0
+    def'.registers
+
+data Access
+  = -- | Managers should only read from this register
+    ReadOnly
+  | -- | Managers should only write to this register
+    WriteOnly
+  | -- | Managers can read from and write to this register
+    ReadWrite
+  deriving (Show)
```

Comment:
Perhaps add a TODO that one day we want to support all of this (see "bus access modes"):

https://corsair.readthedocs.io/en/latest/regmap.html#bit-field